src/Parallafka/MessagesByKey.cs

```diff
@@ -0,0 +1,131 @@
+        /// <returns>True if a nextMessage was found</returns>
+        public bool TryGetNextMessageToHandle(IKafkaMessage<TKey, TValue> message, [NotNullWhen(true)] out IKafkaMessage<TKey, TValue> nextMessage)
+        {
+            lock (this._messagesToHandleForKey)
+            {
+                if (!this._messagesToHandleForKey.TryGetValue(message.Key, out var messagesQueuedForKey))
+                {
+                    // shouldn't happen
+                    nextMessage = null;
+                    Parallafka<TKey, TValue>.WriteLine($"MBK:GetNext: {message.Key} {message.Offset} [none]");
+                    return false;
+                }
+
+                if (messagesQueuedForKey == null || messagesQueuedForKey.Count == 0)
+                {
+                    this._messagesToHandleForKey.Remove(message.Key);
+
+                    if (this._completed && this._messagesToHandleForKey.Count == 0)
+                    {
+                        this._completedSource.TrySetResult();
+                    }
+
+                    nextMessage = null;
+
+                    Parallafka<TKey, TValue>.WriteLine($"MBK:GetNext: {message.Key} {message.Offset} [none]");
+
+                    return false;
+                }
+
+                nextMessage = messagesQueuedForKey.Dequeue();
+
+                Parallafka<TKey, TValue>.WriteLine($"MBK:GetNext: {message.Key} {message.Offset} Next:{nextMessage.Offset}");
+
+                return true;
+            }
+        }
+
+        /// <summary>
+        /// Returns true if the message should be handled, false otherwise
+        /// </summary>
+        /// <param name="message"></param>
+        /// <returns></returns>
+        public bool TryAddMessageToHandle(IKafkaMessage<TKey, TValue> message)
+        {
+            lock (this._messagesToHandleForKey)
+            {
+                var aMessageWithThisKeyIsCurrentlyBeingHandled = this._messagesToHandleForKey.TryGetValue(message.Key, out var messagesToHandleForKey);
+
+                if (aMessageWithThisKeyIsCurrentlyBeingHandled)
+                {
+                    if (messagesToHandleForKey == null)
+                    {
+                        messagesToHandleForKey = new Queue<IKafkaMessage<TKey, TValue>>();
+                        this._messagesToHandleForKey[message.Key] = messagesToHandleForKey;
+                    }
+
+                    Parallafka<TKey, TValue>.WriteLine($"MBK:Enqueuing: {message.Key} {message.Offset}");
+
+                    messagesToHandleForKey.Enqueue(message);
+
+                    return false;
+                }
+
+                // Add the key to indicate that a message with the key is being handled (see above)
+                // so we know to queue up any additional messages with the key.
+                // Without this line, FIFO same-key handling order is not enforced.
+                // Remove it to test the tests.
+                lock (this._messagesToHandleForKey)
```

Comment:
What's up with this lock? Leftover from your initial work? (TIL this apparently doesn't deadlock)

Reply:
This lock isn't needed.  Yes: nested locks in the same thread do not deadlock.  You can create locks that do deadlock in the same thread, but it's a special case when you'd want to do that.